packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,493 @@
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const chainId = parseCaipAssetType(asset);
+      if (!assetsByScope[chainId]) {
+        assetsByScope[chainId] = [];
+      }
+      assetsByScope[chainId].push(asset);
+    }
+
+    let newMetadata: Record<CaipAssetType, AssetMetadata>;
+    for (const chainId in assetsByScope) {
+      if (Object.prototype.hasOwnProperty.call(assetsByScope, chainId)) {
+        const assetsForChain = assetsByScope[chainId as CaipChainId];
+        // Now fetch metadata from the associated asset Snaps:
+        const snap = this.#getAssetSnapFor(chainId as CaipChainId);
+        if (snap) {
+          const metadata = await this.#getMetadata(assetsForChain, snap.id);
+          newMetadata = {
+            ...this.state.metadata,
+            ...metadata,
+          };
+        }
+      }
+    }
+    this.update((state) => {
+      state.metadata = newMetadata;
+    });
+  }
+
+  /**
+   * Creates a mapping of CAIP-2 Chain ID to Asset Snaps.
+   *
+   * @returns A mapping of CAIP-2 Chain ID to Asset Snaps.
+   */
+  #getAssetSnaps(): Record<CaipChainId, Snap[]> {
+    const snaps: Record<CaipChainId, Snap[]> = {};
+    const allSnaps = this.#getAllSnaps();
+    const allPermissions = allSnaps.map((snap) =>
+      this.#getSnapsPermissions(snap.id),
+    );
+    // start mock
+    allPermissions.forEach((singlePermission) => {
+      (singlePermission as unknown as AssetEndowment) = {
+        ...singlePermission,
+        'endowment:assets': {
+          scopes: ['bip122:000000000019d6689c085ae165831e93'],
+        },
+      };
+    });
+    (allPermissions[0] as unknown as AssetEndowment) = {
+      ...allPermissions[0],
+      'endowment:assets': {
+        scopes: ['solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1'],
+      },
+    };
+    // end mock
+
+    for (const [index, permission] of allPermissions.entries() as unknown as [
+      number,
+      AssetEndowment,
+    ][]) {
+      const scopes = permission['endowment:assets']?.scopes;
+      if (!scopes) {
+        continue;
+      }
+      for (const scope of scopes) {
+        if (!snaps[scope]) {
+          snaps[scope] = [];
+        }
+        snaps[scope].push(allSnaps[index]);
+      }
+    }
+    return snaps;
+  }
+
+  /**
+   * Returns the first asset snap for the given scope
+   *
+   * @param scope - The scope to get the asset snap for
+   * @returns The asset snap for the given scope
+   */
+  #getAssetSnapFor(scope: CaipChainId): Snap | undefined {
+    const allSnaps = this.#snaps[scope];
+    // Pick only the first one, we ignore the other Snaps if there are multiple candidates for now.
+    return allSnaps?.[0]; // Will be undefined if there's no Snaps candidate for this scope.
+  }
+
+  /**
+   * Returns all the asset snaps
+   *
+   * @returns All the asset snaps
+   */
+  #getAllSnaps(): Snap[] {
+    return this.messagingSystem.call('SnapController:getAll') as Snap[];
```

Comment:
```suggestion
    return this.messagingSystem.call('SnapController:getAll').filter((snap) => snap.enabled && !snap.blocked);
```

We should probably add a getter for this, will make a note

Reply:
Thanks updated here [6e738c9](https://github.com/MetaMask/core/pull/5138/commits/6e738c94845f2a901914a55cc6443f80df3d3c07)